doc/developer/platform/architecture-storage.md

```diff
@@ -107,25 +107,150 @@ All commands are durably persisted and their effect survives crashes and reboots
+You could certainly hand roll something that also has the properties, but then you have a bunch of work to do to convince folks.
+By following the recipe, the goal is that you shouldn't have to convince anyone of anything.
+
+### Ingredients
+
+1. We have some "raw sources" of data.
+
+    These sources implement something like our `SourceReader` trait, which has a method `get_next_message()` that produces an ordered sequence of messages.
+    Messages come with some ordered "address" (e.g. a (part, offset) pair or a LSN position or byte offset).
+    The messages are emitted in an order that matches the address (the address never goes backwards).
+    The source can be initialized at an address.
+    For each address, the source produces the same message to all readers (or "NA" if it's gone).
+
+2. We have access to the `persist` framework that can record evolving collections
+
+    The framework accepts arbitrary `Append` statements to individual shards and confirm when and whether they are durably recorded.
+    It handles contended writes, picking at most one winner, and providing a single consistent history of updates for each shard.
+    A collection may be formed of multiple shards, which may not have the same durable frontiers.
+
+3. We have an evolving understanding of what should happen to our data from raw sources to `persist`.
+
+    We want some amount of flexibility here, as we may be surprised by the requirements of source decoding, or new opportunities.
+    The "understanding" should reflect deterministic computation.
+    Non-determinism is something we can do with an advanced recipe (more `persist` stuff, in the spirit of reclocking, where we record non-deterministic choices).
+
+### The recipe
+
+There is some prep work to do first:
+
+1.  Identify all "raw sources" that contribute to some intended output (e.g. kafka topics, metadata, schema registry).
+
+2.  "Reclock" each input, by choosing and recording in `persist` shards a collection of source addresses that are monotonically increasing (do not go backwards, but needn't always go forward).
+
+    We should use times that roughly track "real time" so that they align with the times of other collections.
+    The reclocked inputs are going to be treated as if their data "become available" at the first moment for which the data's address is covered by the reclocking collection.
+
+    As suggested, this may include things like the schema registry that aren't conventional inputs, but nonetheless are external sources that influence the output we will produce.
+
+2. Identify the "state" that a streaming implementation of the intended functionality would need to maintain.
+
+    For example, if we are matching up transaction metadata with the data it references, we may need to maintain some outstanding metadata records and some outstanding data records, until all in a transaction match.
```

Comment:
Yes, though I didn't mention it because I was smart but because that text already existed. It is a great argument for why one may not be able to naively reclock everything. Or perhaps that the reclocking should be memoized, and not re-performed for independent sources (if that accomplishes the same thing). 

I was hoping that the *state* would be pipeline local, as it is a deterministic function of the clocked input collections.

Reply:
> I was hoping that the state would be pipeline local, as it is a deterministic function of the clocked input collections.

yes, definitely agree with this!